pydrive2/fs/spec.py

```diff
@@ -534,13 +534,22 @@ def _gdrive_upload_fobj(self, title, parent_id, stream, callback=None):
         return item
 
     def cp_file(self, lpath, rpath, **kwargs):
-        """In-memory streamed copy"""
-        with self.open(lpath) as stream:
-            # IterStream objects doesn't support full-length
-            # seek() calls, so we have to wrap the data with
-            # an external buffer.
-            buffer = io.BytesIO(stream.read())
-            self.upload_fobj(buffer, rpath)
+        """copy a single file or folder"""
+
+        if self.exists(rpath):
+            raise FileExistsError(rpath)
+
+        file_id = self._get_item_id(lpath)
+
+        file = self.client.CreateFile({"id": file_id})
+
+        if file["mimeType"] == FOLDER_MIME_TYPE:
+            dst_id = self._get_item_id(self._parent(rpath))
+            self._gdrive_create_dir(dst_id, rpath.split("/")[-1])
```

Comment:
do you plan to implement `recursive` later here?

Reply:
the recursive part is already in `AbstractFileSystem.copy`:

```python
    def copy(self, path1, path2, recursive=False, on_error=None, **kwargs):
        """Copy within two locations in the filesystem

        on_error : "raise", "ignore"
            If raise, any not-found exceptions will be raised; if ignore any
            not-found exceptions will cause the path to be skipped; defaults to
            raise unless recursive is true, where the default is ignore
        """
        if on_error is None and recursive:
            on_error = "ignore"
        elif on_error is None:
            on_error = "raise"

        paths = self.expand_path(path1, recursive=recursive)
        path2 = other_paths(paths, path2)
        for p1, p2 in zip(paths, path2):
            try:
                self.cp_file(p1, p2, **kwargs)
            except FileNotFoundError:
                if on_error == "raise":
                    raise
```

and it does work with my implementation on `cp_file`

Reply:
make sense ... so, what is the semantics of this block then? could you describe it please?

Reply:
I actually did not find any example of how it's supposed to work, i learned that by trial and error...

What i learnt is:

starting with a structure:

- `self.path = root/tmp`
- `root/tmp/a.pdf` -> regular pdf file
- `root/tmp/folder1/` -> folder with files and sub-folders inside
- `root/tmp/folder/` -> empty folder

the tests i did are:

- `fs.copy('root/tmp/a.pdf', 'root/tmp/folder/b.pdf')` -> will copy and rename the file
- fs.copy('root/tmp/folder1', 'root/tmp/folder/new_folder1', recursive=True) -> will copy and rename the folder, also copy the content of `folder1` into `new_folder1`
- if you copy a file and do not specify the target name it will work:
```python
fs.copy('root/tmp/a.pdf', 'root/tmp/folder/')
# > will create a file in root/tmp/folder named a.pdf
```
- if you copy a folder without specifying the target name it will do something unexpected
```python
fs.copy('root/tmp/folder1/', 'root/tmp/folder/', recursive=True)
# > will create another root/tmp/folder/ and copy the content of folder1 there.
# ( maybe we can fix this one by checking if the folder exist before creating it )
``` 
- if you copy a file with `recursive=True` it does work
- if you copy a folder with `recursive=False` it will just create the folder

@shcheklein do you have some resources to understand how it's supposed to work and if this is the correct implementation?

Reply:
Okay, it makes sense. It's unfortunate that it's called `cp_file` in `fsspec`.

I like the approach of testing `copy` itself to get to the right semantics. I think in this case we can try to get as close as possible to UNIX `cp` semantics.

It means, for example, that `if you copy a folder with recursive=False it will just create the folder` should not work.

Also, I would check in what cases we create parents, we should not be doing this. Even in recursive mode, if `rpath` parent doesn't exist we should not be creating it to my mind.

> if you copy a folder without specifying the target name it will do something unexpected

yep, make sense to fix it also. Probably it should be creating "root/tmp/folder/folder1".